### PR TITLE
Feature/graph view streamlining

### PR DESCRIPTION
**Pull request by ljeub-pometry**
### What changes were proposed in this pull request?

- Factor out internal graph api into a trait that should be implemented on views
- Improvements to the public GraphViewOps api 
- Implement GraphViewOps for Graph and WindowedGraph
- Make all algorithms generic over graph type (no more forcing windows in rust)

### Why are the changes needed?

- Make the overall structure cleaner and make it possible to implement different view structs without breaking all algorithms.

### How was this patch tested?

All tests still work 

### Are there any further changes required?

Python API needs to be revisited to work with different views (it is still only implemented for WindowedGraph)

### CODE REVIEW

File: docbrown/core/src/eval.rs [hunk 1]
```diff
@@ -10,7 +10,7 @@ use std::{
 use rustc_hash::FxHashSet;
 
 use crate::{
-    tgraph::{TemporalGraph, VertexView},
+    tgraph::{TemporalGraph, VertexReference},
```
Comment (fabianmurariu) [inline docbrown/core/src/eval.rs:13]: Should we call it `VRef` or `VertexRef` ?

File: docbrown/db/src/vertex.rs [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+use crate::view_api::internal::GraphViewInternalOps;
+use crate::view_api::{VertexListOps, VertexViewOps};
+use docbrown_core::tgraph::VertexReference;
+use docbrown_core::{Direction, Prop};
+use std::collections::HashMap;
+use std::sync::Arc;
+
+pub struct VertexView<G: GraphViewInternalOps> {
+    //FIXME: Not sure Arc is good here, maybe this should just own a graph and rely on cheap clone...
+    graph: Arc<G>,
```
Comment (fabianmurariu) [inline docbrown/db/src/vertex.rs:11]: Depends what graph do you expect to have here
`Graph` in db is already clone friendly so you can own it, TemporalGraph not so much

File: docbrown/core/src/eval.rs (hunk 1 above)
Comment (ljeub-pometry) [reply docbrown/core/src/eval.rs]: `VertexRef` and `EdgeRef` it is for now

File: docbrown/core/src/tgraph.rs [hunk 3]
```diff
@@ -274,32 +274,32 @@ impl TemporalGraph {
...
                 pid: Some(*pid),
             },
         })
     }
 
-    pub(crate) fn vertex_window(&self, v: u64, w: &Range<i64>) -> Option<VertexView> {
+    pub(crate) fn vertex_window(&self, v: u64, w: &Range<i64>) -> Option<VertexRef> {
         let pid = self.logical_to_physical.get(&v)?;
         let w = w.clone();
         let mut vs = self.index.range(w.clone()).flat_map(|(_, vs)| vs.iter());
```
Comment (fabianmurariu) [inline docbrown/core/src/tgraph.rs:294]: why not kmerge and dedup here?

**Review by fabianmurariu — APPROVED** [review]
LGTM

File: docbrown/db/src/graph_window.rs [hunk 4]
```diff
@@ -48,367 +50,393 @@ pub struct WindowedGraph {
...
+        d: Direction,
+    ) -> Box<dyn Iterator<Item = EdgeRef> + Send> {
+        self.graph
+            .vertex_edges_window_t(v, self.actual_start(t_start), self.actual_end(t_end), d)
     }
 
-    fn out_neighbours(self) -> Self {
-        Box::new(self.flat_map(|v| v.out_neighbours()))
+    fn neighbours(&self, v: VertexRef, d: Direction) -> Box<dyn Iterator<Item = VertexRef> + Send> {
+        self.graph.neighbours_window(v, self.t_start, self.t_end, d)
```
Comment (miratepuffin) [inline docbrown/db/src/graph_window.rs:249]: Once we fix the window bounds, do these two (neighbours and neighbours_window) end up basically with the same semantics? I imagine we need to keep both to make the trait happy?

**Review by miratepuffin — APPROVED** [review]

File: docbrown/core/src/tgraph.rs (hunk 3 above)
Comment (ljeub-pometry) [reply docbrown/core/src/tgraph.rs:294]: I did not write this code but having had a look at it, it is indeed terribly inefficient. However, no need for kmerge and dedup, this should be iterating over the bit sets until it finds one that contains the vertex or there are no more. This is turning that into a linear search for no reason. This should just be using `any` on the range iterator...

Comment (ljeub-pometry) [reply docbrown/core/src/tgraph.rs:294]: I made a ticket to fix this.

File: docbrown/db/src/graph_window.rs (hunk 4 above)
Comment (ljeub-pometry) [reply docbrown/db/src/graph_window.rs:249]: Note that neighbours_window allows you to restrict the view further, may be useful for chaining views...

**Pull request merged**